onnxruntime/core/providers/cuda/tensor/nonzero_impl.cu

```diff
@@ -0,0 +1,116 @@
+{
+  size_t temp_storage_bytes = 0;
+  CUDA_CALL_THROW(cub::DeviceScan::InclusiveSum(
+      nullptr, temp_storage_bytes, prefix_counts, prefix_counts, number_of_blocks));
+  return temp_storage_bytes;
+}
+
+void NonZeroInclusivePrefixSum(void* d_temp_storage, size_t temp_storage_bytes, int* prefix_counts, int number_of_blocks)
+{
+  CUDA_CALL_THROW(cub::DeviceScan::InclusiveSum(
+      d_temp_storage, temp_storage_bytes, prefix_counts, prefix_counts, number_of_blocks));
+}
+
+template<typename InputT, int THREADS_PER_BLOCK>
+__global__ 
+void NonZeroCountEachBlockKernel(const InputT* x, int x_size, int *count_in_blocks)
+{
+  typedef cub::BlockReduce<int, THREADS_PER_BLOCK, cub::BLOCK_REDUCE_RAKING_COMMUTATIVE_ONLY> BlockReduceT;
+  __shared__ typename BlockReduceT::TempStorage temp_storage;
+
+  const cub::CastOp<bool> cast_to_bool;
+  int nz = 0;
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (index < x_size && cast_to_bool(x[index])) ++nz;
+  int count = BlockReduceT(temp_storage).Sum(nz);
+
+  if (threadIdx.x == 0) {
+    count_in_blocks[blockIdx.x] = count;
+  }
+}
+
+template<typename InputT, int THREADS_PER_BLOCK>
+__global__ 
+void NonZeroOutputPositionsKernel(
+    const InputT *x, int x_size, int x_rank, const fast_divmod* x_strides,
+    const int* prefix_counts, int nonzero_elements, int64_t* results)
+{
+  typedef cub::BlockScan<int, THREADS_PER_BLOCK> BlockScanT;
+  __shared__ typename BlockScanT::TempStorage temp_storage;
+
+  const cub::CastOp<bool> cast_to_bool;
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+  int nz = 0;
+  if (index < x_size && cast_to_bool(x[index])) ++nz;
+  int pos_in_block = 0;
+  BlockScanT(temp_storage).InclusiveSum(nz, pos_in_block);
+
+  int result_position = ((blockIdx.x == 0) ? 0 : prefix_counts[blockIdx.x - 1]) + pos_in_block - nz;
+
+  if ((index < x_size) && cast_to_bool(x[index])) {
+    int remain = (int)index, dim = 0;
+    for (int axis = 0, rp = result_position; axis < x_rank; ++axis, rp += nonzero_elements) {
+        x_strides[axis].divmod(remain, dim, remain);
+        results[rp] = (int64_t)dim;
+    }
+  }
+}
+
+template<typename InputT>
+void NonZeroCountEachBlock(const InputT* x, int x_size, int* count_in_blocks)
+{
+  cudaGetLastError();
+  int num_blocks = NonZeroCalcBlockCount(x_size);
+  NonZeroCountEachBlockKernel<InputT, THREADS_PER_BLOCK><<<num_blocks, THREADS_PER_BLOCK>>>(
+      x, x_size, count_in_blocks);
+  CUDA_CALL_THROW(cudaGetLastError());
+}
+    
+template<typename InputT>
+void NonZeroOutputPositions(
+    const InputT *x, int x_size, int x_rank, const fast_divmod* x_strides,
+    const int* prefix_counts, int nonzero_elements, int64_t* results)
+{
+  int num_blocks = NonZeroCalcBlockCount(x_size);
+  NonZeroOutputPositionsKernel<InputT, THREADS_PER_BLOCK><<<num_blocks, THREADS_PER_BLOCK>>>(
+      x, x_size, x_rank, x_strides,
+      prefix_counts, nonzero_elements, results);
+  CUDA_CALL_THROW(cudaGetLastError());
```

Comment:
>cudaGetLastError [](start = 18, length = 16)

remove?

Reply:
I do meet situations during debugging that kernel launch failed when the code is not fully ready. Having this would report error as early as possible. Yet seems most of our calls did not check. Should we enforce the kernel launch checking or not? This is related with above issues.

Reply:
When developing this op, this code might be useful, but do we need to keep it after the implementation is stabilized?

Reply:
Yeah. Removed.